src/main/java/works/crossdock/CrossdockServerInboundHandler.java

```diff
@@ -0,0 +1,135 @@
+  public CrossdockServerInboundHandler(Map<String, Behavior> inputBehaviors) {
+    for (Entry<String, Behavior> entry : inputBehaviors.entrySet()) {
+      behaviors.put(entry.getKey(), entry.getValue());
+    }
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest fullHttpRequest)
+      throws Exception {
+    QueryStringDecoder queryStringDecoder = new QueryStringDecoder(fullHttpRequest.uri());
+    CrossdockRequest request = populateRequest(queryStringDecoder);
+    String behavior = request.getParam("behavior");
+    if (behavior == null) {
+      FullHttpResponse httpResponse =
+          new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK);
+      ctx.writeAndFlush(httpResponse).addListener(ChannelFutureListener.CLOSE);
+      return;
+    }
+    Behavior br = behaviors.get(behavior);
+    runBehavior(br, request)
+        .whenComplete(
+            (response, ex) -> {
+              if (ex != null) {
+                ctx.fireExceptionCaught(ex);
+              } else {
+                writeAndCloseChannel(ctx, response);
+              }
+            });
+  }
+
+  /**
+   * Runs the request against the passed behavior.
+   *
+   * @param br behavior to run against
+   * @param req request to run
+   * @return CompletionStage that contains crossdockResponse
+   * @throws Exception if behavior throws an exception
+   */
+  public CompletionStage<CrossdockResponse> runBehavior(Behavior br, CrossdockRequest req)
+      throws Exception {
+    if (br == null) {
+      return CompletableFuture.completedFuture(
+          new CrossdockResponse().skipped("Unsupported behavior: " + req.getParam("behavior")));
+    }
+
+    return br.run(req)
+        .exceptionally(
+            ex -> {
+              String exceptionAsString = Throwables.getStackTraceAsString(ex);
+              return new CrossdockResponse().error(exceptionAsString);
+            });
+  }
+
+  private void writeAndCloseChannel(ChannelHandlerContext ctx, CrossdockResponse response) {
+    ObjectMapper objectMapper = new ObjectMapper();
+    byte[] body;
+    HttpResponseStatus status = HttpResponseStatus.OK;
+    try {
+      body = objectMapper.writeValueAsBytes(response.getResults());
+    } catch (JsonProcessingException e) {
+      body = e.getMessage().getBytes(Charsets.UTF_8);
+      status = HttpResponseStatus.INTERNAL_SERVER_ERROR;
+    }
+    FullHttpResponse httpResponse =
+        new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, status, Unpooled.wrappedBuffer(body));
+    ctx.writeAndFlush(httpResponse).addListener(ChannelFutureListener.CLOSE);
+  }
+
+  /**
+   * Constructs a crossdockRequest from passed queryStringDecoder.
+   *
+   * @param queryStringDecoder decoder to construct request from
+   * @return crossDockRequest formed from the queryStringDecoder
+   */
+  public CrossdockRequest populateRequest(QueryStringDecoder queryStringDecoder) {
```

Comment:
it doesn't populate and existing object. It decodes to a new one.
`decodeRequest` sounds like a better name.